File: RequestQueue.cpp
#include "RequestQueue.h"
#include <iostream>
using namespace std;

RequestQueue::RequestQueue(int maxSize) 
    : front(nullptr), rear(nullptr), queueSize(0), maxSize(maxSize) {}

RequestQueue::~RequestQueue() {
    clearQueue();
}

void RequestQueue::clearQueue() {
    while (!isEmpty()) {
        QueueNode* temp = front;
        front = front->next;
        // Note: We don't delete request object as it's managed elsewhere
        delete temp;
    }
    rear = nullptr;
    queueSize = 0;
}

bool RequestQueue::enqueue(ParkingRequest* request) {
    if (isFull()) {
        cout << "Error: Request queue is full. Cannot add more requests." << endl;
        return false;
    }
    
    QueueNode* newNode = new QueueNode(request);
    
    if (isEmpty()) {
        front = rear = newNode;
    } else {
        rear->next = newNode;
        rear = newNode;
    }
    
    queueSize++;
    return true;
}

ParkingRequest* RequestQueue::dequeue() {
    if (isEmpty()) {
        return nullptr;
    }
    
    QueueNode* temp = front;
    ParkingRequest* request = temp->request;
    
    front = front->next;
    if (front == nullptr) {
        rear = nullptr;
    }
    
    delete temp;
    queueSize--;
    return request;
}

ParkingRequest* RequestQueue::peek() const {
    if (isEmpty()) {
        return nullptr;
    }
    return front->request;
}

bool RequestQueue::isEmpty() const {
    return front == nullptr;
}

bool RequestQueue::isFull() const {
    return queueSize >= maxSize;
}

int RequestQueue::getSize() const {
    return queueSize;
}

void RequestQueue::displayQueue() const {
    cout << "\n=== PENDING REQUESTS QUEUE ===" << endl;
    cout << "Queue Size: " << queueSize << "/" << maxSize << endl;
    
    if (isEmpty()) {
        cout << "Queue is empty." << endl;
        return;
    }
    
    QueueNode* current = front;
    int counter = 1;
    
    while (current != nullptr) {
        cout << "\n" << counter << ". ";
        cout << "Request ID: " << current->request->getRequestId();
        cout << ", Vehicle: " << current->request->getVehicle()->getVehicleId();
        cout << ", Zone: " << current->request->getRequestedZoneId();
        cout << ", State: " << current->request->stateToString();
        current = current->next;
        counter++;
    }
    cout << endl;
}

void RequestQueue::clear() {
    clearQueue();
}

File: python-flask-version/app/request_queue.py
from .parking_request import ParkingRequest

class RequestQueue:
    class QueueNode:
        def __init__(self, request):
            self.request = request
            self.next = None
    
    def __init__(self, max_size=100):
        self.front = None
        self.rear = None
        self.queue_size = 0
        self.max_size = max_size
    
    def enqueue(self, request):
        """Enqueue a request."""
        if self.is_full():
            print("Error: Request queue is full. Cannot add more requests.")
            return False
        
        new_node = self.QueueNode(request)
        
        if self.is_empty():
            self.front = self.rear = new_node
        else:
            self.rear.next = new_node
            self.rear = new_node
        
        self.queue_size += 1
        return True
    
    def dequeue(self):
        """Dequeue a request."""
        if self.is_empty():
            return None
        
        temp = self.front
        request = temp.request
        
        self.front = self.front.next
        if self.front is None:
            self.rear = None
        
        self.queue_size -= 1
        return request
    
    def peek(self):
        """Peek at the front request."""
        if self.is_empty():
            return None
        return self.front.request
    
    def is_empty(self):
        """Check if queue is empty."""
        return self.front is None
    
    def is_full(self):
        """Check if queue is full."""
        return self.queue_size >= self.max_size
    
    def get_size(self):
        """Get queue size."""
        return self.queue_size
    
    def display_queue(self):
        """Display the queue."""
        print("\n=== PENDING REQUESTS QUEUE ===")
        print(f"Queue Size: {self.queue_size}/{self.max_size}")
        
        if self.is_empty():
            print("Queue is empty.")
            return
        
        current = self.front
        counter = 1
        
        while current is not None:
            print(f"\n{counter}. ", end="")
            print(f"Request ID: {current.request.get_request_id()}", end="")
            print(f", Vehicle: {current.request.get_vehicle().get_vehicle_id()}", end="")
            print(f", Zone: {current.request.get_requested_zone_id()}", end="")
            print(f", State: {current.request.state_to_string()}")
            current = current.next
            counter += 1
    
    def get_all_requests(self):
        """Get all requests as a list."""
        requests = []
        current = self.front
        while current is not None:
            requests.append(current.request)
            current = current.next
        return requests
    
    def clear(self):
        """Clear the queue."""
        while not self.is_empty():
            self.dequeue()